### PR TITLE
[CELEBORN-267] reuse stream when client channel reconnected

### PR DESCRIPTION
**Pull request by RexXiong**
### What changes were proposed in this pull request?
Reuse the stream when channel inactive while fetching trunk.

### Why are the changes needed?
Stream will be removed when channel inactive while fetching trunk, so even if the client is reconnected, fetch trunk would not success and then client will retry other peer reader for same partition location. We can reuse the stream when client is reconnected to avoid this.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
random kill workers & worker high-pressure test

### CODE REVIEW

File: common/src/main/java/org/apache/celeborn/common/network/server/ChunkStreamManager.java [hunk 1]
```diff
@@ -41,14 +40,15 @@ public class ChunkStreamManager {
...
+  // ShuffleKey -> StreamId
+  protected final ConcurrentHashMap<String, Set<Long>> shuffleStreamIds;
 
   /** State of a single stream. */
   protected static class StreamState {
     final FileManagedBuffers buffers;
-
-    // The channel associated to the stream
-    final Channel associatedChannel;
+    final String shuffleKey;
```
Comment (waitinfuture) [inline common/src/main/java/org/apache/celeborn/common/network/server/ChunkStreamManager.java:51]: nice catch!

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala [hunk 2]
```diff
@@ -230,6 +230,7 @@ private[celeborn] class Worker(
     val estimatedAppDiskUsage = new JHashMap[String, JLong]()
     activeShuffleKeys.addAll(partitionLocationInfo.shuffleKeySet)
     activeShuffleKeys.addAll(storageManager.shuffleKeySet())
+    activeShuffleKeys.addAll(fetchHandler.shuffleKeySet())
```
Comment (waitinfuture) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:233]: is this necessary?

File: common/src/main/java/org/apache/celeborn/common/network/server/ChunkStreamManager.java [hunk 3]
```diff
@@ -154,18 +149,47 @@ public long chunksBeingTransferred() {
...
     return myStreamId;
   }
 
+  public Set<String> shuffleKeySet() {
+    return shuffleStreamIds.keySet();
+  }
+
+  public void cleanupExpiredShuffleKey(Set<String> expiredShuffleKeys) {
+    for (String expiredShuffleKey : expiredShuffleKeys) {
+      Set<Long> removedStreamId = shuffleStreamIds.remove(expiredShuffleKey);
```
Comment (waitinfuture) [inline common/src/main/java/org/apache/celeborn/common/network/server/ChunkStreamManager.java:177]: removedStreamId -> expiredStreams

File: common/src/main/java/org/apache/celeborn/common/network/server/ChunkStreamManager.java [hunk 4]
```diff
@@ -41,14 +40,15 @@ public class ChunkStreamManager {
   private static final Logger logger = LoggerFactory.getLogger(ChunkStreamManager.class);
 
   private final AtomicLong nextStreamId;
+  // StreamId -> StreamState
   protected final ConcurrentHashMap<Long, StreamState> streams;
+  // ShuffleKey -> StreamId
+  protected final ConcurrentHashMap<String, Set<Long>> shuffleStreamIds;
```
Comment (waitinfuture) [inline common/src/main/java/org/apache/celeborn/common/network/server/ChunkStreamManager.java:46]: I think we need ConcurrentSet here

Comment (RexXiong) [reply common/src/main/java/org/apache/celeborn/common/network/server/ChunkStreamManager.java:46]: It's already concurrentHashSet as value is ConcurrentHashMap.newKeyset

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks!

**Pull request merged**